runtime/memprof.c

```diff
@@ -289,44 +285,38 @@ static void purge_postponed_queue(void)
    block is allocated, but not yet initialized, so that the heap
    invariants are broken. */
 static void register_postponed_callback(value block, uintnat occurrences,
-                                        enum ml_alloc_kind kind)
+                                        enum ml_alloc_kind kind,
+                                        value* callstack)
 {
-  value callstack;
   struct postponed_block* new_hd;
   if (occurrences == 0) return;
-  callstack = capture_callstack_postponed();
-  if (callstack == 0) return;    /* OOM */
+  if (*callstack == 0) *callstack = capture_callstack_postponed();
+  if (*callstack == 0) return;    /* OOM */
 
-  new_hd = postponed_hd + 1;
-  if (new_hd == postponed_queue_end) new_hd = postponed_queue;
+  new_hd = postponed_next(postponed_hd);
   if (new_hd == postponed_tl) {
```

Comment:
What's the motivation for reimplementing this resize code? The new version's harder to read: the old one with the while loop was obviously correct, and I'm not sure whether the new version is. Is it for performance? How often do you expect this to run?

Reply:
I had the feeling that the `memcpy` code was more natural (eh, we are just moving some data, after all!), and yes, performance is better.

But I agree the end result is not so good. I reverted the initial code.